webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp

```diff
@@ -0,0 +1,128 @@
+using std::placeholders::_2;
+
+namespace webots_ros2_driver {
+  void Ros2Receiver::init(webots_ros2_driver::WebotsNode *node, std::unordered_map<std::string, std::string> &parameters) {
+    Ros2SensorPlugin::init(node, parameters);
+    // This parameter is read when loading the URDF file
+    mDeviceName = parameters.count("name") ? parameters["name"] : "emitter";
+    mDeviceChannel = parameters.count("channel") ? atoi(parameters["channel"].c_str()) : -1;
+
+    mReceiver = wb_robot_get_device(mDeviceName.c_str());
+    mIsEnabled = false;
+
+    assert(mReceiver != 0);
+
+    wb_receiver_set_channel(mReceiver, mDeviceChannel);
+
+    // Data publisher
+    mDataPublisher =
+      mNode->create_publisher<webots_ros2_msgs::msg::StringStamped>(mTopicName + "/data", rclcpp::SensorDataQoS().reliable());
+    // Services
+    enable_service_ = mNode->create_service<webots_ros2_msgs::srv::SetInt>(
+      mTopicName + "/enable", std::bind(&Ros2Receiver::enable_callback, this, _1, _2));
+    get_emitter_direction_service_ = mNode->create_service<webots_ros2_msgs::srv::ReceiverGetEmitterDirection>(
+      mTopicName + "/get_emitter_direction", std::bind(&Ros2Receiver::get_emitter_direction_callback, this, _1, _2));
+    get_sampling_period_service_ = mNode->create_service<webots_ros2_msgs::srv::GetInt>(
+      mTopicName + "/get_sampling_period", std::bind(&Ros2Receiver::get_sampling_period_callback, this, _1, _2));
+    get_signal_strength_service_ = mNode->create_service<webots_ros2_msgs::srv::GetFloat>(
+      mTopicName + "/get_signal_strength", std::bind(&Ros2Receiver::get_signal_strength_callback, this, _1, _2));
+    RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " initialized!").c_str());
+
+    // Calculate timestep
+    if (mAlwaysOn) {
+      wb_receiver_enable(mReceiver, mPublishTimestepSyncedMs);
+      RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " activated!").c_str());
+      mIsEnabled = true;
+    }
+  }
+  void Ros2Receiver::step() {
+    if (!preStep())
+      return;
+
+    if (mIsEnabled)
+      publishData();
+
+    if (mAlwaysOn)
+      return;
+
+    mustPublish();
+  }
+  bool Ros2Receiver::mustPublish() {
+    // Enable/Disable sensor
+    const bool shouldBeEnabled = mDataPublisher->get_subscription_count() > 0;
+    if (shouldBeEnabled != mIsEnabled) {
+      if (shouldBeEnabled)
+        wb_receiver_enable(mReceiver, mPublishTimestepSyncedMs);
+      else
+        wb_receiver_disable(mReceiver);
+      mIsEnabled = shouldBeEnabled;
+    }
+    return mIsEnabled;
+  }
+  void Ros2Receiver::publishData() {
+    // If there is any packet publish the data
+    if (wb_receiver_get_queue_length(mReceiver) > 0) {
+      const int length_buffer = wb_receiver_get_data_size(mReceiver);
+      const void *received_data = wb_receiver_get_data(mReceiver);
+      // cast to char*
+      char *publishing_data_char = (char *)realloc(const_cast<void *>(received_data), length_buffer);
```

Comment:
Why do we need to realloc here? That seems useless. It seems we only need to perform a cast, isn't it?

Reply:
Yes, you are right, there were previous tests that troubled the decoding part but that was only an wrong length value on the `wb_emitted_send()` function from `Ros2Emitter`.

```suggestion
      char *publishing_data_char = static_cast<void *>(const_cast<void *>(received_data));
```